pkg/oauth/apiserver/registry/useroauthaccesstokens/delegate/delegate.go

```diff
@@ -0,0 +1,218 @@
+		return nil, err
+	}
+
+	tokenListWatcher, err := r.accessTokenStorage.Watch(ctx, sanitizedListOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := newOAuthAccessTokenWatcher(tokenListWatcher, ctxUserName)
+	go watcher.Watch(ctx)
+	return watcher, nil
+}
+
+func (r *REST) ConvertToTable(ctx context.Context, object runtime.Object, tableOptions runtime.Object) (*metav1.Table, error) {
+	return r.tableConvertor.ConvertToTable(ctx, object, tableOptions)
+}
+
+func (r *REST) NamespaceScoped() bool {
+	return false
+}
+
+func (r *REST) Delete(ctx context.Context, name string, validateFunc rest.ValidateObjectFunc, options *metav1.DeleteOptions) (runtime.Object, bool, error) {
+	ctxUserName, err := getUserFromContext(ctx)
+	if err != nil {
+		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	var deletedRV string
+	if options.Preconditions != nil && options.Preconditions.ResourceVersion != nil {
+		deletedRV = *options.Preconditions.ResourceVersion
+	}
+
+	deletedObj, err := r.Get(ctx, name, &metav1.GetOptions{TypeMeta: options.TypeMeta, ResourceVersion: deletedRV})
+	if err != nil {
+		return nil, false, err
+	}
+
+	deletedTyped := deletedObj.(*oauthapi.UserOAuthAccessToken)
+	if !isValidUserToken((*oauthapi.OAuthAccessToken)(deletedTyped), ctxUserName) {
+		return nil, false, errors.NewNotFound(oauth.Resource("useroauthaccesstokens"), name)
+	}
+
+	// don't mutate the original options
+	newOpts := options.DeepCopy()
+	if newOpts.Preconditions == nil {
+		newOpts.Preconditions = &metav1.Preconditions{UID: &deletedTyped.UID}
+	} else if newOpts.Preconditions.UID == nil {
+		newOpts.Preconditions.UID = &deletedTyped.UID
+	}
+
+	return r.accessTokenStorage.Delete(ctx, name, validateFunc, newOpts)
+}
+
+func getUserFromContext(ctx context.Context) (string, error) {
+	userInfo, exists := apirequest.UserFrom(ctx)
+	if !exists {
+		klog.V(4).Infof("no user found in the context")
+		return "", kerrors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
+	}
+
+	userName := userInfo.GetName()
+	if len(userName) == 0 {
+		klog.V(4).Info("user in context has zero-length username")
+		return "", kerrors.NewNotFound(oauth.Resource("useroauthaccesstokens"), "")
+	}
+
+	return userName, nil
+}
+
+func oauthAccessTokenListToUserOAuthAccessTokenList(l *oauthapi.OAuthAccessTokenList, username string) *oauthapi.UserOAuthAccessTokenList {
+	ret := oauthapi.UserOAuthAccessTokenList{}
+	for _, t := range l.Items {
+		if !isValidUserToken(&t, username) {
+			continue
+		}
+		ret.Items = append(ret.Items, oauthapi.UserOAuthAccessToken(t))
+	}
+
+	ret.ListMeta = l.ListMeta
+	return &ret
+}
+
+func userFieldSelectorToListOptions(opts *metainternal.ListOptions, userName string) (*metainternal.ListOptions, error) {
+	var newOpts *metainternal.ListOptions
+	if opts == nil {
+		newOpts = &metainternal.ListOptions{}
+	} else {
+		newOpts = opts.DeepCopy()
+	}
+
+	userSelector := fields.Set{"userName": userName}.AsSelector()
```

Comment:
what if `opts` already specifies the field selector for the `userName`?

I think we should be fine since we have an additional check `isValidUserToken` but maybe we could forbid specifying that particular selector?

Reply:
we call `AndSelectors` which will make sure our user is always checked for. If the user checks for another one, the selector will always be false. That's what we want. Compare e2e tests. There is one for that case.